clients/application-kafka/src/main/java/org/eclipse/hono/application/client/kafka/impl/KafkaBasedCommandSender.java

```diff
@@ -268,62 +275,62 @@ void setCorrelationIdSupplier(final Supplier<String> correlationIdSupplier) {
 
     private Handler<DownstreamMessage<KafkaMessageContext>> getCommandResponseHandler(final String tenantId) {
         return message -> {
-            if (message.getCorrelationId() != null) {
+            if (message.getCorrelationId() == null) {
+                LOGGER.trace("ignoring received command response - no correlation id set [tenant: {}]", tenantId);
+                return;
+            }
+            final var pendingResponsesForTenant = pendingCommandResponses.get(tenantId);
+            if (pendingResponsesForTenant != null && !pendingResponsesForTenant.isEmpty()) {
                 // if the correlation id of the received response matches with that of the command
-                Optional.ofNullable(pendingCommandResponses.get(tenantId))
-                        .map(ids -> ids.remove(message.getCorrelationId()))
+                Optional.ofNullable(pendingResponsesForTenant.remove(message.getCorrelationId()))
                         .ifPresentOrElse(
                                 expiringCommandPromise -> expiringCommandPromise
                                         .tryCompleteAndCancelTimer(Future.succeededFuture(message)),
-                                () -> LOGGER.trace(
-                                        "ignoring received command response [correlation-id: {}] as there is no matching correlation id found",
-                                        message.getCorrelationId()));
+                                () -> LOGGER.trace("ignoring received command response - no response pending for correlation id [{}], tenant [{}]",
+                                        message.getCorrelationId(), tenantId));
+            } else {
+                LOGGER.trace("ignoring received command response - no responses pending for tenant [correlation-id: {}, tenant: {}]",
```

Comment:
I think we could leave the code as before(from line 273)  and we could avoid adding the above log statement and we could append the tenant id to the existing log statement and could achieve the same result.

Reply:
My intention here was to explicitly distinguish between "no entry found for correlation id" and "pendingCommandResponses.get(tenantId) is null".
A "pendingCommandResponses.get(tenantId) is null" case log entry would have probably helped in identifying the issue with the recent `KafkaBasedCommandSenderTest` failures.